rust-bins/src/bin/user_cli.rs

```diff
@@ -548,8 +1103,71 @@ fn handle_create_credential(cc: CreateCredential) -> anyhow::Result<()> {
     } else {
         println!(
             "Wrote the credential to {}. Add it to a transaction and send it.",
-            cc.out.display()
+            args.out.display()
         );
     }
     Ok(())
 }
+
+fn handle_recovery(girr: GenerateIdRecoveryRequest) -> anyhow::Result<()> {
+    let ip_info = read_ip_info(&girr.ip_info).context(format!(
+        "Could not read the identity provider information from {}.",
+        girr.ip_info.display()
+    ))?;
+
+    let global_ctx = read_global_context(&girr.global).context(format!(
+        "Could not read cryptographic parameters from {}.",
+        girr.global.display()
+    ))?;
+
+    let use_mainnet = Confirm::new()
+        .with_prompt("Do you want to recover a Mainnet identity? (If not, Testnet will be used)")
+        .interact()
+        .unwrap_or(false);
+
+    let bip39_map = bip39_map();
+
+    let words_str = {
+        println!("Please enter existing phrase below.");
+        let input_words = match read_words_from_terminal(24, true, &bip39_map) {
+            Ok(words) => words,
+            Err(e) => bail!(format!("Error: {}", e)),
+        };
+
+        input_words.join(" ")
+    };
+
+    let wallet = ConcordiumHdWallet {
+        seed: words_to_seed(&words_str),
+        net:  if use_mainnet {
+            Net::Mainnet
+        } else {
+            Net::Testnet
+        },
+    };
+    // We ask what identity and regid index they would like to use.
+    let identity_index = Input::new()
+        .with_prompt("Identity index")
+        .interact()
+        .unwrap_or(0);
+
+    let id_cred_sec_scalar = match wallet.get_id_cred_sec(identity_index) {
+        Ok(scalar) => scalar,
+        Err(e) => {
+            bail!(format!("Could not get idCredSec because {}", e));
+        }
+    };
+
+    let id_cred_sec: PedersenValue<ExampleCurve> = PedersenValue::new(id_cred_sec_scalar);
+    let timestamp = chrono::Utc::now().timestamp() as u64;
+    let request = generate_id_recovery_request(&ip_info, &global_ctx, &id_cred_sec, timestamp);
+    let json = Versioned {
+        version: Version::from(0),
+        value:   request,
+    };
+    write_json_to_file(&girr.request_file, &json).context(format!(
+        "Could not write id recovery request to to {}.",
+        girr.request_file.display()
```

Comment:
Perhaps it would be useful to make it clear at the end here whether mainnet/testnet identity recovery request was generated. It is easy to mistakenly choose the wrong thing in the confirmation dialogue above.